activation/atx_service_db.go

```diff
@@ -0,0 +1,156 @@
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/sql"
+	"github.com/spacemeshos/go-spacemesh/sql/atxs"
+)
+
+// dbAtxService implements AtxService by accessing the state database.
+type dbAtxService struct {
+	golden    types.ATXID
+	logger    *zap.Logger
+	db        sql.Executor
+	atxsdata  *atxsdata.Data
+	validator nipostValidator
+	cfg       dbAtxServiceConfig
+}
+
+type dbAtxServiceConfig struct {
+	// delay before PoST in ATX is considered valid (counting from the time it was received)
+	postValidityDelay time.Duration
+	trusted           []types.NodeID
+}
+
+type dbAtxServiceOption func(*dbAtxServiceConfig)
+
+func WithPostValidityDelay(delay time.Duration) dbAtxServiceOption {
+	return func(cfg *dbAtxServiceConfig) {
+		cfg.postValidityDelay = delay
+	}
+}
+
+func WithTrustedIDs(ids ...types.NodeID) dbAtxServiceOption {
+	return func(cfg *dbAtxServiceConfig) {
+		cfg.trusted = ids
+	}
+}
+
+func NewDBAtxService(
+	db sql.Executor,
+	golden types.ATXID,
+	atxsdata *atxsdata.Data,
+	validator nipostValidator,
+	logger *zap.Logger,
+	opts ...dbAtxServiceOption,
+) *dbAtxService {
+	cfg := dbAtxServiceConfig{
+		postValidityDelay: time.Hour * 12,
+	}
+
+	for _, opt := range opts {
+		opt(&cfg)
+	}
+
+	return &dbAtxService{
+		golden:    golden,
+		logger:    logger,
+		db:        db,
+		atxsdata:  atxsdata,
+		validator: validator,
+		cfg:       cfg,
+	}
+}
+
+func (s *dbAtxService) Atx(_ context.Context, id types.ATXID) (*types.ActivationTx, error) {
+	atx, err := atxs.Get(s.db, id)
+	if errors.Is(err, sql.ErrNotFound) {
+		return nil, ErrNotFound
+	}
+	return atx, err
+}
+
+func (s *dbAtxService) LastATX(ctx context.Context, id types.NodeID) (*types.ActivationTx, error) {
+	atxid, err := atxs.GetLastIDByNodeID(s.db, id)
+	switch {
+	case errors.Is(err, sql.ErrNotFound):
+		return nil, ErrNotFound
+	case err != nil:
+		return nil, fmt.Errorf("getting last ATXID: %w", err)
+	}
+	return atxs.Get(s.db, atxid)
+}
+
+func (s *dbAtxService) PositioningATX(ctx context.Context, maxPublish types.EpochID) (types.ATXID, error) {
+	latestPublished, err := atxs.LatestEpoch(s.db)
+	if err != nil {
+		return types.EmptyATXID, fmt.Errorf("get latest epoch: %w", err)
+	}
+	s.logger.Info("searching for positioning atx", zap.Uint32("latest_epoch", latestPublished.Uint32()))
+
+	// positioning ATX publish epoch must be lower than the publish epoch of built ATX
+	positioningAtxPublished := min(latestPublished, maxPublish)
+	id, err := findFullyValidHighTickAtx(
+		ctx,
+		s.atxsdata,
+		positioningAtxPublished,
+		s.golden,
+		s.validator, s.logger,
+		VerifyChainOpts.AssumeValidBefore(time.Now().Add(-s.cfg.postValidityDelay)),
+		VerifyChainOpts.WithTrustedIDs(s.cfg.trusted...),
+		VerifyChainOpts.WithLogger(s.logger),
+	)
+	if err != nil {
+		s.logger.Info("search failed - using golden atx as positioning atx", zap.Error(err))
+		id = s.golden
+	}
```

Comment:
I think in this case rather than returning the golden ATX it should return `types.EmptyATXID` and let the caller decide how to proceed (the caller knows the golden ATX but will most likely chose `PrevATX` instead (unless its an initial ATX)